README.pod
=pod

README for PDL::SVDLIBC

=head1 ABSTRACT

PDL::SVDLIBC - PDL interface to Doug Rohde's SVD C Library

=head1 REQUIREMENTS

=over 4

=item * PDL

Tested version 2.4.2.

=item SVDLIBC

By Dough Rohde, based on the SVDPACKC library,
which was written by Michael Berry, Theresa Do,
Gavin O'Brien, Vijay Krishna and Sowmini Varadhan.
Tested version 1.33.

Available from http://tedlab.mit.edu/~dr/SVDLIBC/

=back


=head1 DESCRIPTION

PDL::SVDLIBC provides a PDL interface to the SVDLIBC routines
for singular value decomposition of large sparse matrices.

=head1 BUILDING

Build this module as you would any perl module, by doing something
akin to the following:

 gzip -dc distname-XX.YY.tar.gz | tar -xof -
 cd distname-XX.YY/
 perl Makefile.PL
 make
 make test                                     # optional
 make install

See L<perlmodinstall>(1) for details.

During the build process, you may be prompted for the locations
of required libraries, header files, etc.

=head1 AUTHOR

Bryan Jurish E<lt>jurish@ling.uni-potsdam.deE<gt>

=head1 COPYRIGHT

Copyright (c) 2005, Bryan Jurish.  All rights reserved.

This package is free software, and entirely without warranty.
You may redistribute it and/or modify it under the same terms
as Perl itself.

=cut


ChangeLog
ChangeLog for perl/PDL module PDL::SVDLIBC

v0.16 Tue, 24 Nov 2015 09:44:58 +0100 moocow
	+ cpantesters win32 builds now failing with "svdutil.o:svdutil.c:(.text+0xabd): undefined reference to `htonl'"
	+ replacing netinet/in.h htonl() and ntohl() calls with dummy macros, since they're not really used here

v0.15 Thu, 19 Nov 2015 14:28:15 +0100 moocow
	+ still no joy with v0.14 cpan smoker hacks on win32
	  - see http://www.cpantesters.org/cpan/report/6e49170d-6bf5-1014-a25c-daa0d353e6bb (UNKNOWN)
	+ packaging patched SVDLIBC sources (BSD-like license) into perl distro
	+ added svdindexND(), svdindexNDt(), svdindexccs() for fast decoding of large sparse SVD-encoded matrices
	+ added svderror() for precision checking (slow but memory-friendly)

v0.14 Tue, 17 Nov 2015 14:52:16 +0100 moocow
	+ v0.13 inet.h patch strategy seems still not to be catching
	  - see e.g. http://www.cpantesters.org/cpan/report/6c6dcdff-6c03-1014-8ce0-5f22b8efd73c
	  - test disabling Config.cache and local SVDLIBC source directories in Config.PL
	+ added whichND()-style interface svdlas2w(), svdlas2aw()
	  - should be slightly less memory-intensive than pdl-side ccsencode() + svdlas2()
	+ added tests for convenience wrappers and reduced svd (d<n)

v0.13 Mon, 03 Aug 2015 09:22:03 +0200 moocow
	+ patch upstream svdutil.c to disable #include <netinet/inet.h>
	  - allowcpantesters win32 builds
	  - it's totally unclear why <inet.h> is included in the first place

v0.12 Sat, 01 Aug 2015 12:02:13 +0200 moocow
	+ Config.PL don't quote typename macros (e.g. -D__SVDLIBC_LONG='PDL_Long'), for cpantesters/win32

v0.11 Thu, 30 Jul 2015 09:37:16 +0200 moocow
	+ Config.PL set binmode() on archive filehandle for LWP::UserAgent method (for win32)

v0.10 Tue, 28 Jul 2015 09:13:34 +0200 moocow
	+ Makefile calls to hack-int-types.perl tweaked for BSD variants
	  - some cpantesters builds failed with 'Using $< in a non-suffix rule context is a GNUmake idiom (Makefile:367)'
	  - see http://cpantesters.org/cpan/report/18c0b8a6-3263-11e5-a85a-e28fcaadd3a7
	+ unlink any stale sources when unpacking upstream archive in Config.PL
	+ disabled lwp-request fetch method in Config.PL (binary redirection doesn't work for win32)

v0.09 Fri, 24 Jul 2015 12:38:03 +0200 moocow
	+ added las2.patch for upstream SVDLIBC sources
	  - disable implicit matrix transposition in svdLAS2() at SVDLIBC/las2.c:373-378

v0.08 Thu, 26 Sep 2013 12:13:31 +0200 moocow
	+ added auto-download and -unpack of SVDLIBC source to Config.PL
	+ use ExtUtils::MakeMaker::prompt() for user prompts (fix RT #88972)

v0.07 Tue, 24 Sep 2013 14:30:46 +0200 moocow
	+ PREREQ_PM{PDL}=CONFIGURE_REQUIRES{PDL}=0 for cpan-testers friendliness
	+ added (DISABLED) support for PDL_Indx datatype (PDL >= v2.007)
	  - disabled by default since which(), whichND(), & co. don't return PDL_Indx in devel PDL-2.006_90
	  - to enable, set $cconfig{ENABLE_PDL_INDX}=1 in Config.PL, run make realclean && perl Makefile.PL && ...

v0.06 2013-06-19  moocow
	+ include $Config{ccflags} explicitly to avoid errors like
	  "t/00_basic.t .. Not a CODE reference at /usr/lib/perl/5.14/DynaLoader.pm line 207."
	+ added pdlmaker.plm for CPAN indexing

v0.05 2009-10-30  moocow
	+ removed DEBUG flags
	+ should now compile & run on both 32- and 64-bit

v0.04 Tue, 27 Feb 2007 18:58:12 +0100
	+ replaced PDL 'longlong' specs with 'int', jives again with SVDLIBC 'long'
	+ there's got to be a better way to do this...

v0.01 Mon, 20 Jun 2005 23:59:20 +0200
	+ initial version


Makefile.PL
use PDL::Core::Dev;
use ExtUtils::MakeMaker;

##-- read in user variables
do "Config.PL";
die "$0: reading 'Config.PL' failed: $@" if ($@);

##-- check for required SVDLIBC source directory
foreach (map {"$cconfig{SVD_SRCDIR}/$_"} qw(svdutil.h svdutil.c svdlib.h svdlib.c las2.c)) {
  if (!-e $_) {
    warn "$0: could not find required SVDLIBC source file '$_', bailing out";
    exit 0; ##-- exit with 0 status to appease cpan-testers (avoid FAIL reports)
  }
}

##-- pdlmaker stuff
PDL::Core::Dev->import();
require "pdlmaker.plm";

$package = ["SVDLIBC.pd",SVDLIBC,PDL::SVDLIBC];
%hash = pdlmaker_init($package);

$hash{AUTHOR} = 'Bryan Jurish';
$hash{ABSTRACT} = 'PDL interface to Doug Rohde\'s SVD C Library';
$hash{PREREQ_PM} = {
		    PDL=>0,
		   };
$hash{CONFIGURE_REQUIRES} = {
			     'PDL'=>0,
			     'ExtUtils::MakeMaker'=>0,
			     'Data::Dumper'=>0,
			    };
$hash{DIR} = [];
$hash{VERSION_FROM} = 'SVDLIBC.pd';
$hash{clean}{FILES} .= 'las2.c svdlib.c svdutil.c svdutil.h svdlib.h';
$hash{realclean}{FILES} .= '*~ *.tmp README.txt Config.cache';

push(@{$hash{LIBS}},    ' '.$cconfig{SVD_LIBS});
$hash{INC}           .= ' '.$cconfig{SVD_INC};
$hash{LDFLAGS}       .= ' '.$cconfig{SVD_LDFLAGS};
$hash{OPTIMIZE}      .= ' '.$cconfig{SVD_CFLAGS};

##-- include $Config{ccflags} explicitly to avoid errors like:
## t/00_basic.t .. Not a CODE reference at /usr/lib/perl/5.14/DynaLoader.pm line 207.
use Config;
$hash{CCFLAGS}       .= " $Config{ccflags} ".$cconfig{SVD_DEFS};
$hash{OBJECT}        .= ' las2.o svdlib.o svdutil.o';


WriteMakefile(%hash);

sub MY::depend {
  package MY;
  my $inherited = shift->SUPER::depend(@_);
  $inherited .= (
		 "\n\n"
		 ##--
		 ."SVDLIBC.o: svdutil.h svdlib.h\n\n"
		 ."las2.o: las2.c svdutil.h svdlib.h\n\n"
		 ."svdlib.o: svdlib.c svdutil.h svdlib.h\n\n"
		 ."svdutil.o: svdutil.c svdutil.h svdlib.h\n\n"
		);
}

sub MY::special_targets {
  package MY;
  my $inherited = shift->SUPER::special_targets(@_);
  #$inherited =~ s/^(\.SUFFIXES\s*:.*)$/$1 .pod .rpod .pm .man .txt .html;/
  $inherited .= "

##-- SVDLIBC sources
SVDLIBC = $main::cconfig{SVD_SRCDIR}
".join('',
       map {
	 "
$_: \$(SVDLIBC)/$_
\t\$(PERL) hack-int-types.perl < \$(SVDLIBC)/$_ > \$@

##-- readme
readme: README.txt

"
       } qw(las2.c svdlib.c svdlib.h svdutil.c svdutil.h));
}

##-- avoid applying 'processPL' rules to 'Config.PL'
sub MY::processPL { return ''; }


Config.PL
## File: PDL-SVDLIBC/Config.PL
## Description: user variables for PDL::SVDLIBC package

##-- load cache (disabling to see if this fixes cpantesters failures (v0.14, Tue, 17 Nov 2015 14:56:15 +0100)
if (0 && -e "./Config.cache") {
  do "./Config.cache";
  if ($@) {
    warn("$0: could not load cache data from 'Config.cache': $@");
  }
}

##--
## $val = cprompt($key, $message)
## $val = cprompt($key, $message, $default)
##   + sets $cconfig{$key}
sub cprompt {
  my ($key, $msg, $default)=@_;
  return $cconfig{$key} if (defined($cconfig{$key}));
  $default = '' if (!defined($default)); 
  my $answer = ExtUtils::MakeMaker::prompt("<Config.PL> $msg [$default] ?");
  chomp($answer);
  return $cconfig{$key} = ($answer eq '' ? $default : $answer);
}

##--------------------------------------------------------------
## Sources

## $SVD_SRCDIR : location of SVDLIBC sources
#my $SVD_SRC0 = (grep {-d $_} qw(./SVDLIBC ../SVDLIBC))[0] // "http://tedlab.mit.edu/~dr/SVDLIBC/svdlibc.tgz";
my $SVD_SRCDIR  = 'SVDLIBC';
my $SVD_SRCFILE = 'svdlibc.tgz';
my $SVD_SRC0 = (
		#(grep {-d $_} qw(./SVDLIBC ../SVDLIBC)), ##-- disable in-place SRCDIR to see if that helps cpantesters failures (v0.14, Tue, 17 Nov 2015 14:55:43 +0100)
		#(grep {-d $_} qw(../SVDLIBC)), ##-- debug
		(grep {-f $_} ($SVD_SRCFILE, 'svdlibc.tar.gz')),
		"http://tedlab.mit.edu/~dr/SVDLIBC/svdlibc.tgz",
	       )[0];
my $SVD_SRC  = cprompt('SVD_SRC', "SVDLIBC source directory, archive, or URL", $SVD_SRC0);

$SVD_SRC =~ s{^file://}{};
if (-d $SVD_SRC) {
  ##-- local directory source
  $cconfig{SVD_SRCDIR} = $SVD_SRC;
}
elsif (-f $SVD_SRC) {
  ##-- local file source: buffer
  ;
}
elsif ($SVD_SRC =~ m{^\w+://}) {
  ##-- url source
  print STDERR "Info: attempting to download SVDLIBC source from $SVD_SRC\n";
  $SVD_SRC = get_url($SVD_SRC,$SVD_SRCFILE);
}

if (-f $SVD_SRC) {
  ##-- source archive: unpack it
  print STDERR "Info: attempting to unpack SVDLIBC source from $SVD_SRCFILE ... ";
  require Archive::Tar;
  require File::Basename;
  my $tar = Archive::Tar->new($SVD_SRCFILE)
    or die("Error: SVDLIBC source file $SVD_SRCFILE doesn't look like a tar archvie");
  $tar->contains_file('SVDLIBC/svdlib.h')
    or die("Error: SVDLIBC source archive $SVD_SRCFILE doesn't contain required file(s)");
  foreach (grep {-f $_ || -l $_} $tar->list_files) {
    unlink($_)
      or die("Warning: failed to remove stale upstream source file $_");
  }
  $tar->extract()
    or die("Error: failed to extract SVDLIBC source from $SVD_SRCFILE");
  print STDERR "success.\n";

  ##-- patch upstream sources
  print STDERR "Info: attempting to patch upstream SVDLIBC sources ...\n";
  my $rc = system("patch -p0 -i svdlibc.patch -N");
  if ($rc==0) {
    print STDERR "Info: patch succeeded.\n";
  } else {
    warn("Warning: failed to patch upstream SVDLIBC sources: segfaults may result");
  }

  $cconfig{SVD_SRCDIR} = './SVDLIBC';

}

##-- sanity check
if (! -d $cconfig{SVD_SRCDIR}) {
  warn("Warning: source directory SVD_SRCDIR=$SVD_SRCDIR does not exist!");
  return;
}


##--------------------------------------------------------------
## Flags

## $SVD_INC : include flags for SVDLIBC
#cprompt('SVD_INC', "include flags for SVDLIBC", '-I/usr/local/include/svdlibc');
#cprompt('SVD_INC', "Additional include flags for SVDLIBC", "-I$cconfig{SVD_SRCDIR}");
#$cconfig{SVD_INC} = "-I$cconfig{SVD_SRCDIR}";
$cconfig{SVD_INC} = "-I.";

## $SVD_DEFS : additional defines for SVDLIBC
cprompt("ENABLE_PDL_INDX", "Enable 64-bit PDL_Indx support if available (requires PDL >= v2.007)", "y");
$cconfig{ENABLE_PDL_INDX} = ($cconfig{ENABLE_PDL_INDX} && $cconfig{ENABLE_PDL_INDX} !~ /^[n0]/ ? 1 : 0);

require PDL::Core;
my ($LONG,$LONGLONG) = $cconfig{ENABLE_PDL_INDX} && defined(&PDL::indx) ? ('PDL_Indx','PDL_Indx') : ('PDL_Long','PDL_LongLong');
cprompt('SVD_LONG',     "SVDLIBC long integer datatype", "$LONG");
cprompt('SVD_LONGLONG', "SVDLIBC long long integer datatype", "$LONGLONG");
cprompt('SVD_DEFS', "Additional defines for SVDLIBC", "-D__SVDLIBC_LONG=$cconfig{SVD_LONG} -D__SVDLIBC_LONG_LONG=$cconfig{SVD_LONGLONG}");

## $SVD_CFLAGS : additional compiler flags for SVDLIBC
# + do *NOT* use -ffast-math !
#my $SVD_CFLAGS = $ENV{CFLAGS};
#my $SVD_CFLAGS = "-O2 -pipe";
#my $SVD_CFLAGS = "-O3 -pipe -march=native -mtune=native -mfpmath=sse";
#cprompt('SVD_CFLAGS', "SVDLIBC compiler flags", $SVD_CFLAGS);

## $SVD_LDFLAGS : linker flags for SVDLIBC
#cprompt('SVD_LDFLAGS', "SVDLIBC linker flags", '-L/usr/local/lib');

## $SVD_LIBS : libs to link
#cprompt('SVD_LIBS', "SVDLIBC libraries", '');

## $SVD_OBJ : additional object files and/or static libs to link
#my $SVD_OBJ = '/usr/local/lib/libsvd.a';
#my $SVD_OBJ = join(' ', qw(las2.o svdlib.o svdutil.o);
#cprompt('SVD_OBJ', "SVDLIBC objects", $SVD_OBJ);

##--------------------------------------------------------------
## save cache file
use Data::Dumper;
open(CACHE,">Config.cache")
  or die("$0: open failed for Config.cache: $!");
print CACHE
  ("## Automatically generated, remove to re-configure!\n",
   Data::Dumper->Dump([\%cconfig],['*cconfig']),
   "1; ##-- be happy\n",
  );
close CACHE;

##--------------------------------------------------------------
## subs

## $outfile_or_undef = get_url($url, $outfile)
##  + URL download code adapted from HTTP::UA::Parser module's Makefile.PL
sub get_url {
  my ($url,$outfile) = @_;

  ##-- curl
  print STDERR "+ Trying curl... ";
  if (0 && system('curl','-sS','-o',$outfile,$url)==0) {
    print STDERR "success.\n";
    return $outfile;
  }
  print STDERR "failed.\n";

  ##-- wget
  print STDERR "+ Trying wget... ";
  if (0 && system('wget', '-nv', '-O',$outfile, $url)==0) {
    print STDERR "success.\n";
    return $outfile;
  }
  print STDERR "failed.\n";

  ##-- try LWP::UserAgent
  print STDERR "+ Trying LWP::UserAgent... ";
  my $rc = 'failed';
  eval "use LWP::UserAgent";
  if ($@) {
    print STDERR
      ("failed.\n",
       "\n",
       " Couldn't locate LWP::UserAgent module\n",
       " LWP::UserAgent is required to fetch SVDLIBC sources from server\n",
       " Please install it or get SVDLIBC sources manually from\n",
       " $url\n",
       " and place it as $SVD_SRCFILE the root folder of this distro\n",
       " then run Makefile.PL again\n",
       "\n",
      );
    exit 1;
  }
  my $ua = LWP::UserAgent->new;
  $ua->timeout(5);
  $ua->env_proxy();
  my $response = $ua->get($url);
  if ($response->is_success) {
    print STDERR "success.\n";
    open(my $outfh,">$outfile")
      or die("Error: open failed for $outfile: $!");
    binmode($outfh); ##-- we might be win32
    print $outfh $response->content;
    close($outfh)
      or die("Error: close failed for $outfile: $!");
    return $outfile;
  }
  else {
    die "LWP::UserAgent request aborted\n";
    exit 1;
  }
}


1; ##-- return nicely

